File: evtq/evtq.go
// Package evtq creates and manages event queues
//
// It depends upon [container/heap] to manage a heap structure
package evtq

import (
	"container/heap"
	"github.com/iti/evt/vrtime"
	"sync"
)

// InvalidEventID will never be returned from
// EventQueue.Insert()
const InvalidEventID = 0

// EventQueue represents the queue
type EventQueue struct {
	evtID    int           // monotonically increasing counter used for default secondary time in event Time
	itemHeap *itemHeapType // data structure holding items, see struct definition for item and itemHeapType
	lookup   map[int]*item // event identifier to event, used for marking events to be ignored
	MaxTime  vrtime.Time   // Largest vrtime.Time value pushed onto to the heap as yet
	mu       sync.Mutex    // used to support thread safety
}

// New is a constructor. Initializes an empty slice of events
func New() *EventQueue {
	return &EventQueue{
		evtID:    InvalidEventID,      // has to have an event id, so include an invalid one at initialization
		itemHeap: &itemHeapType{},     // event list is initialized to be empty of events
		lookup:   make(map[int]*item)} // map to support deletion of events is initially empty
}

// Len returns the number of elements in the queue.
func (p *EventQueue) Len() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	rtn := p.itemHeap.Len()
	return rtn
}

// MinTime returns the Time associated with the next event.
func (p *EventQueue) MinTime() vrtime.Time {
	p.mu.Lock()
	defer p.mu.Unlock()
	rtn := (*p.itemHeap)[0].Time
	return rtn
}

// Insert inserts a new element into the queue. No action is performed on duplicate elements.
func (p *EventQueue) Insert(v any, time vrtime.Time) int {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.evtID++

	// update maximum time of inserted event
	if p.MaxTime.LT(time) {
		p.MaxTime = time
	}

	// if the priority in the time stamp is -1
	// change it to be something monotonically increasing
	// to try to give some determinism for events with the
	// same tick count when some other priority isn't given
	//
	if time.Pri() == -1 {
		time.SetPri(int64(p.evtID))
	}

	// create an item for insertion
	newItem := &item{
		itemID: p.evtID, // identifier for this event
		Value:  v,       // notice that v can be anything, what matters for ordering is time value
		Time:   time}

	heap.Push(p.itemHeap, newItem)
	p.lookup[p.evtID] = newItem
	rtn := p.evtID
	return rtn
}

// Pop removes the element with the least time from the queue and returns it.
// In case of an empty queue, an error is returned.
func (p *EventQueue) Pop() any {
	p.mu.Lock()
	defer p.mu.Unlock()

	popped := heap.Pop(p.itemHeap).(*item)
	delete(p.lookup, popped.itemID)
	rtn := popped.Value
	return rtn
}

// UpdateTime changes the priority of a given item.
// If the specified item is not present in the queue, no action is performed.
func (p *EventQueue) UpdateTime(evtID int, newTime vrtime.Time) {
	p.mu.Lock()
	defer p.mu.Unlock()
	item, present := p.lookup[evtID]

	if !present {
		return
	}

	item.Time = newTime
	heap.Fix(p.itemHeap, item.index)
}

func (p *EventQueue) GetItem(evtID int) any {
	p.mu.Lock()
	defer p.mu.Unlock()
	_, present := p.lookup[evtID]
	if !present {
		return nil
	}
	return p.lookup[evtID]
}


// Remove an element. Returns true on success.
func (p *EventQueue) Remove(evtID int) bool {
	p.mu.Lock()
	defer p.mu.Unlock()
	element, present := p.lookup[evtID]
	if !present {
		return false
	}

	// we're going to push the element to be rid of to the top
	element.Time = vrtime.ZeroTime()
	heap.Fix(p.itemHeap, element.index)

	// pop it off
	popped := heap.Pop(p.itemHeap).(*item)
	delete(p.lookup, popped.itemID)
	return true
}

// itemHeapType is the type of the data structure written to satisfy the Heap interface
type itemHeapType []*item

// The item struct defines the item organized by Time value
type item struct {
	itemID int         // unique identifier with every event inserted into the queue
	Value  any         // completely general payload for the item
	Time   vrtime.Time // the field used to order the elements
	index  int         // the position of the item in the (heap-organized) slice of events
	Cancel bool        // has been marked for removal 
}

// Len, Less, Swap, Push, and Pop are funcs required for a
// a type to satisfy the Heap interface

// Len returns the number of items in the slice of events
func (ih *itemHeapType) Len() int {
	return len(*ih)
}

// Less with arguments i, j returns true if the item in the priority queue in position i
// has an earlier time-stamp than the item in position j
func (ih *itemHeapType) Less(i, j int) bool {
	return (*ih)[i].Time.LT((*ih)[j].Time)
}

// Swap with arguments i, j exchanges the items in positions i and j
func (ih *itemHeapType) Swap(i, j int) {
	(*ih)[i], (*ih)[j] = (*ih)[j], (*ih)[i]
	(*ih)[i].index = i
	(*ih)[j].index = j
}

// Push puts the item given as an argument at the end of the slice of events
// This is not necessarily its final position in the heap
func (ih *itemHeapType) Push(x any) {
	it := x.(*item)
	it.index = len(*ih)
	*ih = append(*ih, it)
}

// Pop removes the item in the heap which appears in the last position of the heap
func (ih *itemHeapType) Pop() any {
	old := *ih
	item := old[len(old)-1]
	*ih = old[0 : len(old)-1]
	return item
}

File: evtq/evtq.py

# evtq.py: Python translation of Go evtq.go
# Event queue implementation using heapq and threading for thread safety
#
# Package evtq creates and manages event queues
# It depends upon Python's heapq to manage a heap structure

import heapq
import threading
from typing import Any, Dict, Optional
import vrtime


# InvalidEventID will never be returned from EventQueue.Insert()
InvalidEventID = 0

class Item:
    """
    The item struct defines the item organized by Time value
    itemID: unique identifier with every event inserted into the queue
    Value: completely general payload for the item
    Time: the field used to order the elements
    index: the position of the item in the (heap-organized) list of events
    Cancel: has been marked for removal
    """
    def __init__(self, itemID: int, value: Any, time: 'vrtime.Time'):
        self.itemID = itemID
        self.Value = value
        self.Time = time
        self.index = 0
        self.Cancel = False

    def __lt__(self, other: 'Item'):
        return self.Time.LT(other.Time)

class ItemHeap:
    """
    ItemHeap is the type of the data structure written to satisfy the Heap interface
    Implements Len, Less, Swap, Push, and Pop required for a type to satisfy the Heap interface
    """
    def __init__(self):
        self.data = []

    def __len__(self):
        return len(self.data)

    def push(self, item: Item):
        heapq.heappush(self.data, item)

    def pop(self) -> Item:
        return heapq.heappop(self.data)

    def fix(self, index: int):
        heapq.heapify(self.data)

    def get(self, index: int) -> Item:
        return self.data[index]

    def swap(self, i: int, j: int):
        self.data[i], self.data[j] = self.data[j], self.data[i]
        self.data[i].index = i
        self.data[j].index = j

class EventQueue:
    """
    EventQueue represents the queue
    evtID: monotonically increasing counter used for default secondary time in event Time
    itemHeap: data structure holding items
    lookup: event identifier to event, used for marking events to be ignored
    MaxTime: Largest vrtime.Time value pushed onto to the heap as yet
    mu: used to support thread safety
    """
    def __init__(self):
        self.evtID = InvalidEventID
        self.itemHeap = ItemHeap()
        self.lookup: Dict[int, Item] = {}
        self.MaxTime = vrtime.zero_time()
        self.mu = threading.Lock()

    @staticmethod
    def New():
        """New is a constructor. Initializes an empty list of events"""
        return EventQueue()

    def Len(self) -> int:
        """Len returns the number of elements in the queue."""
        with self.mu:
            return len(self.itemHeap)

    def MinTime(self) -> 'vrtime.Time':
        """MinTime returns the Time associated with the next event."""
        with self.mu:
            if len(self.itemHeap) == 0:
                return vrtime.zero_time()
            return self.itemHeap.get(0).Time

    def Insert(self, v: Any, time: 'vrtime.Time') -> int:
        """Insert inserts a new element into the queue. No action is performed on duplicate elements."""
        with self.mu:
            self.evtID += 1

            # update maximum time of inserted event
            if self.MaxTime.LT(time):
                self.MaxTime = time

            # if the priority in the time stamp is -1
            # change it to be something monotonically increasing
            # to try to give some determinism for events with the
            # same tick count when some other priority isn't given
            if time.Pri() == -1:
                time.SetPri(int(self.evtID))

            # create an item for insertion
            newItem = Item(self.evtID, v, time)

            self.itemHeap.push(newItem)
            self.lookup[self.evtID] = newItem
            return self.evtID

    def Pop(self) -> Optional[Any]:
        """Pop removes the element with the least time from the queue and returns it. In case of an empty queue, returns None."""
        with self.mu:
            if len(self.itemHeap) == 0:
                return None
            popped = self.itemHeap.pop()
            self.lookup.pop(popped.itemID, None)
            return popped.Value

    def UpdateTime(self, evtID: int, newTime: 'vrtime.Time'):
        """UpdateTime changes the priority of a given item. If the specified item is not present in the queue, no action is performed."""
        with self.mu:
            item = self.lookup.get(evtID)

            if item is None:
                return
            
            item.Time = newTime
            self.itemHeap.fix(item.index)

    def GetItem(self, evtID: int) -> Optional[Item]:
        """GetItem returns the item for the given event ID, or None if not present."""
        with self.mu:
            return self.lookup.get(evtID)

    def Remove(self, evtID: int) -> bool:
        """Remove an element. Returns True on success."""
        with self.mu:
            element = self.lookup.get(evtID)
            if element is None:
                return False
            
            # we're going to push the element to be rid of to the top
            element.Time = vrtime.zero_time()
            self.itemHeap.fix(element.index)

            # pop it off
            popped = self.itemHeap.pop()
            self.lookup.pop(popped.itemID, None)
            
            return True
